pkg/reconciler/v1alpha1/store/store.go

```diff
@@ -0,0 +1,138 @@
+	Kind           = "Store"
+)
+
+//go:generate counterfeiter . StoreReader
+type StoreReader interface {
+	Read(keychain authn.Keychain, storeImages []expv1alpha1.StoreImage) ([]expv1alpha1.StoreBuildpack, error)
+}
+
+func NewController(opt reconciler.Options, storeInformer v1alpha1expInformers.StoreInformer, storeReader StoreReader, keychainFactory registry.KeychainFactory) *controller.Impl {
+	c := &Reconciler{
+		Client:          opt.Client,
+		StoreLister:     storeInformer.Lister(),
+		StoreReader:     storeReader,
+		KeychainFactory: keychainFactory,
+	}
+	impl := controller.NewImpl(c, opt.Logger, ReconcilerName)
+	storeInformer.Informer().AddEventHandler(reconciler.Handler(impl.Enqueue))
+	return impl
+}
+
+type Reconciler struct {
+	Client          versioned.Interface
+	StoreReader     StoreReader
+	KeychainFactory registry.KeychainFactory
+	StoreLister     v1alpha1expListers.StoreLister
+}
+
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	_, storeName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	store, err := c.StoreLister.Get(storeName)
+	if k8serrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	store = store.DeepCopy()
+
+	store, err = c.reconcileStoreStatus(store)
+
+	updateErr := c.updateStoreStatus(store)
+	if updateErr != nil {
+		return updateErr
+	}
+
+	if err != nil {
+		return controller.NewPermanentError(err)
+	}
+	return nil
+}
+
+func (c *Reconciler) updateStoreStatus(desired *expv1alpha1.Store) error {
+	desired.Status.ObservedGeneration = desired.Generation
+
+	original, err := c.StoreLister.Get(desired.Name)
+	if err != nil {
+		return err
+	}
+
+	if equality.Semantic.DeepEqual(desired.Status, original.Status) {
+		return nil
+	}
+
+	_, err = c.Client.ExperimentalV1alpha1().Stores().UpdateStatus(desired)
+	return err
+}
+
+func (c *Reconciler) reconcileStoreStatus(store *expv1alpha1.Store) (*expv1alpha1.Store, error) {
+	buildpacks, err := c.getBuildpacks(store.Spec.Sources)
+	if err != nil {
+		store.Status = expv1alpha1.StoreStatus{
+			Status: duckv1alpha1.Status{
+				ObservedGeneration: store.Generation,
+				Conditions: duckv1alpha1.Conditions{
+					{
+						Type:               duckv1alpha1.ConditionReady,
+						Status:             corev1.ConditionFalse,
+						LastTransitionTime: apis.VolatileTime{Inner: v1.Now()},
+						Message:            err.Error(),
+					},
+				},
+			},
+		}
+		return store, err
+	}
+
+	store.Status = expv1alpha1.StoreStatus{
+		Buildpacks: buildpacks,
+		Status: duckv1alpha1.Status{
+			ObservedGeneration: store.Generation,
+			Conditions: duckv1alpha1.Conditions{
+				{
+					LastTransitionTime: apis.VolatileTime{Inner: v1.Now()},
+					Type:               duckv1alpha1.ConditionReady,
+					Status:             corev1.ConditionTrue,
+				},
+			},
+		},
+	}
+	return store, nil
+}
+
+func (c *Reconciler) getBuildpacks(sources []expv1alpha1.StoreImage) ([]expv1alpha1.StoreBuildpack, error) {
```

Comment:
this is a stretch goal but I wonder if during reconciliation of the spec, we only fetch buildpacks from images that have been newly added to the spec and not for every image that is mentioned in the spec. 

what I am suggesting is definitely an optimization but it might be an interesting to consider

Reply:
An extension of that would be that when an source image is removed from the spec, we just drop all the buildpacks that were from that image in the store status.

Reply:
I am not sure about the cost/reward ratio there. Lets chat.